mistralrs-core/src/kbnf.rs

```diff
@@ -0,0 +1,103 @@
+use ahash::AHashMap;
+use anyhow::Result;
+use candle_core::{DType, Device, Tensor};
+use kbnf::{
+    engine_like::{AcceptTokenError, MaskLogitsError},
+    AcceptTokenResult, Engine, EngineLike, Token, Vocabulary,
+};
+use tokenizers::Tokenizer;
+
+pub struct KbnfGrammar {
+    engine: Engine,
+    vocab_size: usize,
+}
+
+pub enum KbnfGrammarBias {
+    /// Token was accepted, it can be added to the sequence. No need for resampling.
+    Accepted,
+    /// Token was rejected. Resample with these new logits.
+    /// The token sampled with the bias can be added to the sequence.
+    Resample { new_logits: Tensor },
+    /// Generation was finished, the token can be added to the sequence,
+    /// but no more generation is necessary.
+    FinishedGeneration,
+}
+
+impl KbnfGrammar {
+    pub fn new(grammar: &str, tokenizer: &Tokenizer) -> Result<Self> {
+        let tokenizer_vocab = tokenizer.get_vocab(true);
+        let mut id_to_tok = AHashMap::new();
+        let mut id_to_tok_str = AHashMap::new();
+        for (tok_str, id) in tokenizer_vocab {
+            id_to_tok.insert(id, Token(tok_str.as_bytes().to_vec().into_boxed_slice()));
+            id_to_tok_str.insert(id, tok_str);
+        }
+        let vocab = Vocabulary::new(id_to_tok, id_to_tok_str)?;
+        Ok(Self {
+            engine: Engine::new(grammar, vocab)?,
+            vocab_size: tokenizer.get_vocab_size(true),
+        })
+    }
+
+    /// Compute the bias if this token were to be added.
+    /// If the token can be added
+    pub fn compute_bias_for(
+        &mut self,
+        tok: u32,
+        logits: &Tensor,
+        add_to_trie: bool,
+    ) -> candle_core::Result<KbnfGrammarBias> {
+        // Try to accept the new token
+        match self.engine.try_accept_new_token(tok) {
+            Ok(AcceptTokenResult::Ongoing) => {
+                // Token was accepted, no resampling needed
+                if add_to_trie {
+                    self.engine.compute_allowed_token_ids();
+                }
+                Ok(KbnfGrammarBias::Accepted)
+            }
+            Err(AcceptTokenError::Rejected) => {
+                if add_to_trie {
+                    self.engine.compute_allowed_token_ids();
+                }
+                let mut bias = vec![0f32; self.vocab_size];
+                match self.engine.mask_logits(&mut bias) {
+                    Ok(()) => {
+                        let new_logits = (logits.to_device(&Device::Cpu)?.to_dtype(DType::F32)?
```

Comment:
It might not be a good idea to move logits back to CPU due to potential CUDA synchronization issues. I think we can benchmark this to check latency. If it does bring too much latency, we can follow a strategy like [this file](https://github.com/Dan-wanna-M/kbnf/blob/master/python/kbnf/engine.py) which essentially only moves the indices tensor(tends to be smaller than the whole logits) to GPU and operate there.